src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -773,6 +772,16 @@ PeerImp::doAccept()
         return buf;
     }();
 
+    {
+        // Put a pseudo-message into the queue to hold any messages until we get
+        // the response. This message should never be sent, but create a ping
+        // just in case it "escapes".
+        protocol::TMPing message;
+        message.set_type(protocol::TMPing::ptPING);
+        message.set_seq(0);
+
+        send_queue_.push(std::make_shared<Message>(message, protocol::mtPING));
```

Comment:
This is going to block sending permanently. You need to add `onWriteMessage()` (via strand) to the call-back handler in `async_write()` below. I think we simply should put all initial sends into `doProtocolStart()`. After all that's what the meaning of `doProtocolStart()` is - a peer is ready to send and receive messages. Something like this: https://gist.github.com/gregtatcam/37d5a8c9ae1418828d0f0cf112c83159

Reply:
I agree with @gregtatcam. I have a slightly different implementation, which is along the lines of what Greg has above: https://github.com/nbougalis/rippled/commit/e284d0a18e50b2d486db9841c286e01a51997c89

Generally, I think the protocol _should_ make it very clear than after the onus is on the peer that **established** the connection (i.e. the peer that made the connection which will have `inbound_ == false`): it sends the HTTP upgrade request, and after it receives the response, it must "kick" things by sending further data.

Reply:
@gregtatcam Yep. You're right. My local node ran ok since I started testing this, but any incoming connections lock up until they time out.

@nbougalis So it looks like you moved the `setTimer` and shard request to `doProtocolStart`, which makes sense, and inlined `doAccept` with no significant changes, which seems unnecessary, and IMHO makes `run` harder to read. I pushed a commit which I think works well.